librz/core/cvfile.c

```diff
@@ -0,0 +1,168 @@
+// SPDX-FileCopyrightText: 2021 Florian Märkl <info€florianmaerkl.de>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_core.h>
+
+#define URI_SCHEME     "vfile://"
+#define URI_SCHEME_LEN 8
+
+typedef struct {
+	RzBinFile *bf;
+	RzBinVirtualFile *vfile;
+	ut64 off;
+} VFileCtx;
+
+extern RzIOPlugin rz_core_io_plugin_vfile;
+
+static bool __check(RzIO *io, const char *pathname, bool many) {
+	return !strncmp(pathname, URI_SCHEME, URI_SCHEME_LEN);
+}
+
+static RzIODesc *__open(RzIO *io, const char *pathname, int rw, int mode) {
+	if (strncmp(pathname, URI_SCHEME, URI_SCHEME_LEN)) {
+		return NULL;
+	}
+	char *vfile_path = strdup(pathname + URI_SCHEME_LEN); // path like "<binfile id>/<filename>"
+	if (!vfile_path) {
+		return NULL;
+	}
+	RzIODesc *desc = NULL;
+	char *filename = strchr(vfile_path, '/');
+	if (!filename) {
+		RZ_LOG_ERROR("Invalid URI \"%s\", expected " URI_SCHEME "<fd>/<filename>\n", pathname);
+		goto beach;
+	}
+	*filename++ = '\0';
+	ut32 bfid = (ut32)strtoull(vfile_path, NULL, 0);
+	RzCore *core = io->corebind.core; // We are in the core already so this is fine.
+	RzBinFile *bf = rz_bin_file_find_by_id(core->bin, bfid);
+	if (!bf) {
+		RZ_LOG_ERROR("No bin file for id %" PFMT32u " from URI \"%s\"\n", bfid, pathname);
+		goto beach;
+	}
+	RzBinVirtualFile *vfile = NULL;
+	if (bf->o && bf->o->vfiles) {
+		RzListIter *it;
+		rz_list_foreach (bf->o->vfiles, it, vfile) {
+			if (!strcmp(vfile->name, filename)) {
+				goto found_file;
```

Comment:
Abusing `goto` i'd say :D

Extract a `find_vfile` function and use that instead of multiple gotos.

Reply:
```basic
10 print "fixed"
20 goto 10
```